src/components/userpanel/JumboStampSystem.tsx

```diff
@@ -0,0 +1,632 @@
+
+  &:before {
+    position: absolute;
+    content: "";
+    top: 0;
+    right: 0;
+    bottom: 0;
+    left: 0;
+    z-index: 0;
+    background-color: ${(props: JumboStampSystemProps) =>
+      props.hide ? "#ffc13a" : "#efa74c"};
+    clip-path: ${(props: JumboStampSystemProps) =>
+      props.hide
+        ? "polygon(0% 0,100% 0%,100% 100%,100% 100%,0% 100%)"
+        : "polygon(0% 0,85% 0%,100% 85%,100% 100%,0% 100%)"};
+    transition: all 0.5s ease-out;
+  }
+
+  &:after {
+    position: absolute;
+    content: "";
+    top: 0;
+    right: 0;
+    bottom: 0;
+    left: 0;
+    z-index: 0;
+    background-color: #ffc13a;
+    clip-path: ${(props: JumboStampSystemProps) =>
+      props.hide
+        ? "polygon(0% 0,85% 0%,100% 100%,100% 100%,0% 100%)"
+        : "polygon(0% 0,74% 0%,90% 100%,75% 100%,0% 100%)"};
+    transition: clip-path 0.25s ease;
+  }
+`;
+
+const ArrowDecoration = styled.img`
+  position: relative;
+  z-index: 2;
+  height: 100%;
+  transform: ${(props: JumboStampSystemProps) =>
+    props.active ? "translate(10%, 0)" : "translate(0px, 0)"};
+  transform: ${(props: JumboStampSystemProps) =>
+    props.hide ? "translate(5%) scale(1.6)" : ""};
+  transition: all ease 1.25s;
+  animation: ${(props: JumboStampSystemProps) =>
+    props.hide
+      ? css`
+          ${moveIn} 1.6s cubic-bezier(1,0,0,1)
+        `
+      : css`
+          ${moveInBack} 1.6s cubic-bezier(1,0,0,1)
+        `};
+`;
+
+const RewardsHeaderContainer = styled.div`
+  background-color: #8397cf;
+  height: 4.05vh;
+  clip-path: var(--notched-sm-tp);
+  display: flex;
+  justify-content: flex-end;
+  padding: 1rem;
+`;
+
+const RewardsButtonContainer = styled.div`
+  display: flex;
+  justify-content: flex-end;
+`;
+
+const RewardsButtonInner = styled.div`
+  padding: 1rem 1rem 1rem 0;
+  width: fit-content;
+`;
+
+const RewardsHeader = styled.p`
+  font-size: 2.15vh;
+  color: white;
+  text-overflow: clip;
+  white-space: nowrap;
+  font-family: video-cond, serif;
+  font-weight: 300;
+`;
+
+const RewardsDate = styled.span`
+  padding: 0rem 1rem 0rem 1rem;
+  font-size: 2.25rem;
+  color: #e0ce83;
+  letter-spacing: -1px;
+`;
+
+const MainContainer = styled.div`
+  /* display: none; */
+  animation: ${(props: JumboStampSystemProps) =>
+    props.active
+      ? css`
+          ${changeView} 1s cubic-bezier(1,0,1,-0.07)
+        `
+      : css``};
+  transition: visibility ease 2s;
+  animation-fill-mode: forwards;
+  visibility: ${(props: JumboStampSystemProps) =>
+    props.active ? "hidden" : "visible"};
+  /* display: ${(props: JumboStampSystemProps) =>
+    props.active ? "none" : ""}; */
+`;
+
+interface JumboStampSystemProps {
+  active?: boolean;
+  hide?: boolean;
+}
+
+const JumboStampSystem = () => {
+  const dispatch = useDispatch();
+  const showing = useSelector(selectShowingStamp);
+
+  const [hoverActive, setHoverActive] = useState(false);
+
+  const { account } = useEthers();
+
+  function getStamps(): number[] | undefined {
```

Comment:
It seems like we have this type of function in a few places. I think instead we should create a hook `useStamps` that is responsible for this. This hook will have a state `const [stamps, setStamps] = useState<StampType[] | null>(null);`

Then we will have this `getStamps` function that sets the stamps just once, called in a `useEffect` on first load. We can move all this logic for getting the stamps into this hook. Or maybe this `useEffect` should also update if the `account` updates, so that should be a dependency in this `useEffect`.

Then, whenever we need the stamps in any of our components, we can get the data by just adding:

`const stamps = useStamps();`

And it should make a lot of these components a lot cleaner and simpler.

Reply:
✔️ Done.
Added useStamps in functions.ts:

```javascript
export const useStamps = (
  provider: ethers.providers.Provider | undefined,
  chainId: number
) => {
  const [stamps, setStamps] = useState<StampType[] | null>(null);
  const { account } = useEthers();

  useEffect(() => {
    const getStamps = async () => {
      try {
        if (!provider || !account) return;

        const correctProvider = new ethers.providers.JsonRpcProvider(
          "https://polygon.llamarpc.com"
        );

        const contract = new ethers.Contract(
          CONTRACT_ADDR,
          abiStamps,
          correctProvider
        );
        const ids = Array.from({ length: 12 }, (_, i) => i + 1);
        const results = await contract.balanceOfBatch(
          Array(ids.length).fill(account), // array of addresses
          ids // array of ids
        );
        const parsedResult = results.map((result: BigNumber) => Number(result));
        setStamps(parsedResult);
      } catch (error) {
        console.error("Failed to fetch stamps", error);
      }
    };

    getStamps();
  }, [provider, account, chainId]); // Update stamps whenever the account, provider, or chainId changes

  return stamps;
};
```